src/Clay/Grid.hs

```diff
@@ -1,58 +1,271 @@
+
+-------------------------------------------------------------------------------
+
+-- | Property defines the line names and track sizing functions of the grid rows.
+gridTemplateRows :: GridTrackList a -> Css
+gridTemplateRows = key "grid-template-rows"
 
 -- | Property defines the line names and track sizing functions of the grid columns.
-gridTemplateColumns :: [Size a] -> Css
-gridTemplateColumns = key "grid-template-columns" . noCommas
+gridTemplateColumns :: GridTrackList a -> Css
+gridTemplateColumns = key "grid-template-columns"
+
+newtype GridTrackList a = GridTrackList Value
+  deriving (Val, None, Inherit, Initial, Unset)
+
+instance IsList (GridTrackList a) where
+  type Item (GridTrackList a) = Size a
+  toList = error ""
+  fromList = GridTrackList . noCommas
+
+-------------------------------------------------------------------------------
+
+-- | Property defines the line names and track sizing functions of the grid rows.
+gridAutoRows :: GridAutoTrackList a -> Css
+gridAutoRows = key "grid-auto-rows"
+
+-- | Property defines the line names and track sizing functions of the grid columns.
+gridAutoColumns :: GridAutoTrackList a -> Css
+gridAutoColumns = key "grid-auto-columns"
+
+newtype GridAutoTrackList a = GridAutoTrackList Value
+  deriving (Val, Auto, MinContent, MaxContent, Inherit, Initial, Unset)
+
+instance IsList (GridAutoTrackList a) where
+  type Item (GridAutoTrackList a) = Size a
+  toList = error ""
+  fromList = GridAutoTrackList . noCommas
+
+-------------------------------------------------------------------------------
+gridAutoFlow :: GridAutoFlow -> Css
+gridAutoFlow = key "grid-auto-flow"
+
+newtype GridAutoFlow = GridAutoFlow Value
+  deriving (Val, Row, Column, Inherit, Initial, Unset)
+
+dense :: GridAutoFlow
+dense = GridAutoFlow "dense"
+
+rowDense :: GridAutoFlow
+rowDense = GridAutoFlow "row dense"
+
+columnDense :: GridAutoFlow
+columnDense = GridAutoFlow "column dense"
+-------------------------------------------------------------------------------
+-- | Property defines the element location inside grid template
+gridArea :: GridArea -> Css
+gridArea = key "grid-area"
+
+blankGridArea :: GridArea
+blankGridArea = GridArea "."
+
+newtype GridArea = GridArea Text
+  deriving (IsString, Val)
+
+-------------------------------------------------------------------------------
+
+gridRowStart :: GridLocation -> Css
+gridRowStart = key "grid-row-start"
+
+gridRowEnd :: GridLocation -> Css
+gridRowEnd = key "grid-row-end"
+
+gridColumnStart :: GridLocation -> Css
+gridColumnStart = key "grid-column-start"
+
+gridColumnEnd :: GridLocation -> Css
+gridColumnEnd = key "grid-column-end"
+
+gridLocation :: IsSpan -> These Integer GridArea -> GridLocation
+gridLocation isSpan = GridLocation_Data . GridLocationData isSpan
+
+data IsSpan = Span | NoSpan
+  deriving (Show, Eq)
+
+data GridLocation
+  = GridLocation_Keyword Value
+  | GridLocation_Data GridLocationData
```

Comment:
Let us try and rethink the naming here. Having nearly the same constructor names and type names is a source of confusion.